File: services/frontend-leptos/shared-types/src/voice.rs
use serde::{Deserialize, Serialize};
use crate::guild::GuildVoiceEntry;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct VoiceStatus {
    pub connected: bool,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub active_guild_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub active_channel_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub active_channel_name: Option<String>,
    pub connections: Vec<GuildVoiceEntry>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ActiveSpeaker {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub id: Option<String>,
    pub user_id: String,
    pub username: String,
    pub avatar: String,
    pub speaking: bool,
}

File: services/frontend/src/lib/types/voice.ts
export interface GuildVoiceEntry {
  guildId: string;
  channelId: string;
  channelName: string;
  connectedAt: number;
}

export interface VoiceStatus {
  connected: boolean;
  activeGuildId?: string | null;
  activeChannelId?: string | null;
  activeChannelName?: string | null;
  connections: GuildVoiceEntry[];
  /**
   * Authoritative shared voice snapshot — who is present / speaking right
   * now, aggregated server-side from the gateway's `voice_active_user`
   * deltas. All browsers converge on this same list.
   */
  activeSpeakers?: ActiveSpeaker[];
}

export interface ActiveSpeaker {
  userId: string;
  username: string;
  avatar?: string | null;
  speaking: boolean;
  /** Epoch ms of most recent activity. Stale speakers are auto-expired. */
  lastActiveAt?: number;
}
